src/java/com/unifina/signalpath/utils/MessageChainUtil.java

```diff
@@ -0,0 +1,73 @@
+package com.unifina.signalpath.utils;
+
+import com.streamr.client.protocol.message_layer.MessageID;
+import com.streamr.client.protocol.message_layer.MessageRef;
+import com.streamr.client.protocol.message_layer.StreamMessage;
+import com.streamr.client.protocol.message_layer.StreamMessageV30;
+import com.unifina.data.StreamPartitioner;
+import com.unifina.domain.data.Stream;
+import com.unifina.domain.security.SecUser;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MessageChainUtil implements Serializable {
+	private Map<String,Long> previousTimestamps = new HashMap<>();
+	private Map<String,Long> previousSequenceNumbers = new HashMap<>();
+
+	private transient SecUser cachedUser = null;
+
+	private static final Logger log = Logger.getLogger(MessageChainUtil.class);
+
+	private SecUser getUser(Long userId) {
+		if (cachedUser == null) {
+			cachedUser = SecUser.getViaJava(userId);
+		}
+		return cachedUser;
+	}
+
+	private long getNextSequenceNumber(String key, long timestamp) {
+		if (!previousTimestamps.containsKey(key) || previousTimestamps.get(key) != timestamp) {
+			return 0L;
+		}
+		if (!previousSequenceNumbers.containsKey(key)) {
+			previousSequenceNumbers.put(key, 0L);
+		}
+		return previousSequenceNumbers.get(key)+1;
+	}
+
+	private MessageRef getPreviousMessageRef(String key) {
+		if (!previousTimestamps.containsKey(key)) {
+			return null;
+		}
+		if (!previousSequenceNumbers.containsKey(key)) {
+			previousSequenceNumbers.put(key, 0L);
```

Comment:
- Would be great not to write anything in method `getXXX` (it's not needed here?)
- Is it a valid situation that `previousTimestamp` is found but `previousSequenceNumber` is not? Maybe we should require both to be non-null?

```
Long previousTimestamp = previousTimestamps.get(key);
Long previousSequenceNumber = previousSequenceNumber.get(key);
if (previousTimestamp == null || previousSequenceNumber == null) {
    return null;
} else {
    return new MessageRef(previousTimestamp, previousSequenceNumber);
}